README.md
# Storybook Chrome Screenshot Addon

[storybook]: https://github.com/storybooks/storybook
[puppeteer]: https://github.com/GoogleChrome/puppeteer

![DEMO](https://raw.githubusercontent.com/tsuyoshiwada/storybook-chrome-screenshot/artwork/demo.gif)

[![npm](https://img.shields.io/npm/v/storybook-chrome-screenshot.svg?style=flat-square)](https://www.npmjs.com/package/storybook-chrome-screenshot)
[![CircleCI](https://img.shields.io/circleci/project/github/tsuyoshiwada/storybook-chrome-screenshot/master.svg?style=flat-square)](https://circleci.com/gh/tsuyoshiwada/storybook-chrome-screenshot)
[![David](https://img.shields.io/david/tsuyoshiwada/storybook-chrome-screenshot.svg?style=flat-square)](https://david-dm.org/tsuyoshiwada/storybook-chrome-screenshot)

> A [Storybook][storybook] Addon, Save the screenshot image of your stories :camera: via [Puppeteer][puppeteer].

`storybook-chrome-screenshot` takes a screenshot and saves it.  
It is primarily responsible for image generation necessary for Visual Testing such as `reg-viz`.

## Table of Contents

- [Features](#features)
- [How it works](#how-it-works)
- [Getting Started](#getting-started)
  - [Installation](#installation)
  - [Register Addon](#register-addon)
  - [Register initialization process](#register-initialization-process)
  - [Setup your stories](#setup-your-stories)
    - [React](#react)
    - [Angular](#angular)
    - [Vue.js](#vuejs)
  - [Run `storybook-chrome-screenshot` Command](#run-storybook-chrome-screenshot-command)
  - [Support for addDecorator](#support-for-adddecorator)
- [API](#api)
  - [initScreenshot()](#initscreenshot)
  - [withScreenshot(options = {})](#withscreenshotoptions--)
  - [setScreenshotOptions(options = {})](#setscreenshotoptionsoptions--)
  - [getScreenshotOptions()](#getscreenshotoptions)
- [Command Line Options](#command-line-options)
- [Tips](#tips)
  - [Disable component animation](#disable-component-animation)
- [Examples](#examples)
- [TODO](#todo)
- [Contribute](#contribute)
  - [Development](#development)
    - [`npm run test`](#npm-run-test)
    - [`npm run build`](#npm-run-build)
- [License](#license)

## Features

- :camera: Take screenshots of each stories. via [Puppeteer][puppeteer].
- :rocket: Provide flexible screenshot shooting options.
- :tada: Supports the following framework / View framework.
  - [React](https://github.com/facebook/react/)
  - [Angular](https://github.com/angular/angular)
  - [Vue.js](https://github.com/vuejs/vue)

## How it works

`storybook-chrome-screenshot` executes [Storybook][storybook] in a child process and accesses the launched page using [Puppeteer][puppeteer]. It is a very simple mechanism.  
For that reason, you can easily shoot screenshots by simply creating a story that works with the browser.

## Getting Started

It is very easy to introduce `storybook-chrome-screenshot` in your project.

### Installation

First install `storybook-chrome-screenshot`.

```bash
$ npm install --save-dev storybook-chrome-screenshot
```

> **Note:** Please do not use globally but let it operate locally.

### Register Addon

Next, register Addon.

**.storybook/addons.js**

```javascript
// Other addons...
import 'storybook-chrome-screenshot/register';
```

### Register initialization process

Add [initScreenshot](#initscreenshot) decorator. It has to be **before** the first [withScreenshot](#withscreenshotoptions--) decorator. Addon uses it to catch the finish of the components' rendering.

**Example: .storybook/config.js**

```javascript
import { addDecorator } from '@storybook/react';
import { initScreenshot } from 'storybook-chrome-screenshot';

addDecorator(initScreenshot());
```

### Setup your stories

Create a story with [withScreenshot](#withscreenshotoptions--).

#### React

```javascript
import React from 'react';
import { storiesOf } from '@storybook/react';
import { withScreenshot } from 'storybook-chrome-screenshot';
import Button from './Button';

storiesOf('Button', module).add('with text', withScreenshot()(() => <Button>Text</Button>));
```

#### Angular

This function works well even if you use Angular:

```javascript
import { storiesOf } from '@storybook/angular';
import { withScreenshot } from 'storybook-chrome-screenshot';
import { MyButtonComponent } from '../src/app/my-button/my-button.component';

storiesOf('Button', module).add(
  'with custom label',
  withScreenshot()(() => ({
    component: MyButtonComponent,
    props: {
      text: 'Text'
    }
  }))
);
```

#### Vue.js

Of course, Vue.js works the same way:

```javascript
import { storiesOf } from '@storybook/vue';
import { withScreenshot } from 'storybook-chrome-screenshot';
import MyButton from './Button.vue';

storiesOf('MyButton', module)
  .add(
    'pre-registered component',
    withScreenshot()(() => ({
      template: '<my-button :rounded="true">A Button with rounded edges</my-button>'
    }))
  )
  .add(
    'template + component',
    withScreenshot()(() => ({
      components: { MyButton },
      template: '<my-button>Button rendered in a template</my-button>'
    }))
  )
  .add(
    'render + component',
    withScreenshot()(() => ({
      render: (h) => h(MyButton, { props: { color: 'pink' } }, ['renders component: MyButton'])
    }))
  );
```

### Run `storybook-chrome-screenshot` Command

Open `package.json` and add a `screenshot` script for run `storybook-chrome-screenshot` command.

```json
{
  "scripts": {
    "screenshot": "storybook-chrome-screenshot -p 9001 -c .storybook"
  }
}
```

> **Note:** Parameters such as ports and configuration files should match the parameters of the `Storybook` you are currently using.

After that, just run the `npm run screenshot` command, shotting a component wrapped with [withScreenshot](#withscreenshotoptions--) and save the images.

```bash
$ npm run screenshot
```

### Support for addDecorator

Or by using `addDecorator()`, it is possible to shotting all the decorated stories.

```javascript
import { storiesOf } from '@storybook/react';
import { withScreenshot } from 'storybook-chrome-screenshot';

storiesOf('Button', module)
  .addDecorator(
    withScreenshot({
      /* ...options */
    })
  )
  .add('with primary', () => <Button primary>Primary Button</Button>)
  .add('with secondary', () => <Button secondary>Secondary Button</Button>);
```

## API

### initScreenshot()

This decorator has to be added to every story. Addon uses it to understand when story's rendering is finished.

**Important!.** `initScreenshot` has to be added before the first [withScreenshot](#withscreenshotoptions--).

**Example: .storybook/config.js**

```javascript
import { addDecorator } from '@storybook/react';
import { initScreenshot } from 'storybook-chrome-screenshot';

addDecorator(initScreenshot());
```

### withScreenshot(options = {})

Notify [Puppeteer][puppeteer] of the story wrapped in this function and let it recognize it as the target of the screenshot.

The following objects of `options` can be specified.

```javascript
{
  namespace: 'global',    // namespace for your screenshots. It is using in the filenames, e.g.  Button-with-primary_global.png
  delay: 0,               // Delay milliseconds when shooting screenshots
  waitFor: '',            // User defined trigger function name to shoot screenshots. See "Full control the screenshot timing" section below.
  viewport: {             // Browser's viewport when shooting screenshots. (See: https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pagesetviewportviewport)
    width: 1024,
    height: 768,
    deviceScaleFactor: 1,
    isMobile: false,
    hasTouch: false,
    isLandscape: false,
  },
  filePattern: '{kind}-{story}-{knobs}_{ns}-{vp}' // file pattern, combined from kind, story name, used knobs, namespace and viewport
}
```

Also, By passing the `array` to `viewport`, you can easily shoot multiple Viewports.

```javascript
{
  viewport: [
    // Mobile
    {
      width: 300,
      height: 420,
      isMobile: true,
      hasTouch: true,
    },
    // Tablet
    {
      width: 768,
      height: 800,
      isMobile: true,
      hasTouch: true,
    },
    // Desktop
    {
      width: 1024,
      height: 768,
    },
  ],
}
```

### setScreenshotOptions(options = {})

Sets the default value of the option used with [withScreenshot()](#withscreenshotoptions--).  
It is useful for changing Viewport of all stories.

**Example: .storybook/config.js**

```javascript
import { setScreenshotOptions } from 'storybook-chrome-screenshot';

setScreenshotOptions({
  viewport: {
    width: 768,
    height: 400,
    deviceScaleFactor: 2
  }
});
```

### getScreenshotOptions()

Get the current option used with [withScreenshot()](#withscreenshotoptions--).

```javascript
import { getScreenshotOptions } from 'storybook-chrome-screenshot';

console.log(getScreenshotOptions());
// => Current screenshot options...
```

## Command Line Options

```bash
$ $(npm bin)/storybook-chrome-screenshot --help

  Usage: storybook-chrome-screenshot [options]


  Options:

    -V, --version                    output the version number
    -p, --port [number]              Storybook server port. (default: 9001)
    -h, --host [string]              Storybook server host. (default: localhost)
    -s, --static-dir <dir-names>     Directory where to load static files from.
    -c, --config-dir [dir-name]      Directory where to load Storybook configurations from. (default: .storybook)
    -o, --output-dir [dir-name]      Directory where screenshot images are saved. (default: __screenshots__)
    --parallel [number]              Number of Page Instances of Puppeteer to be activated when shooting screenshots. (default: 4)
    --filter-kind [regexp]           Filter of kind with RegExp. (example: "Button$")
    --filter-story [regexp]          Filter of story with RegExp. (example: "^with\s.+$")
    --inject-files <file-names>      Path to the JavaScript file to be injected into frame. (default: )
    --browser-timeout [number]       Timeout milliseconds when Puppeteer opens Storybook. (default: 30000)
    --puppeteer-launch-config [json] JSON string of launch config for Puppeteer. (default: {"args":["--no-sandbox","--disable-setuid-sandbox", "--disable-dev-shm-usage"]})
    --silent                         Suppress standard output.
    --debug                          Enable debug mode.
    -h, --help                       output usage information
```

## Tips

### Disable component animation

When shooting screenshots, you may want to disable component animation. In this case it is easiest to inject Script using the `--inject-files` option.

You can create `./disable-animation.js` and disable CSS Animation with the next snippet.

```javascript
(() => {
  const $iframe = document.getElementById('storybook-preview-iframe');
  const $doc = $iframe.contentDocument;
  const $style = $doc.createElement('style');

  $style.innerHTML = `* {
    transition: none !important;
    animation: none !important;
  }`;

  $doc.body.appendChild($style);
})();
```

Pass the created file to the `--inject-files` option.

```bash
$ $(npm bin)/storybook-chrome-screenshot --inject-files ./disable-animation.js [...more options]
```

### Full control the screenshot timing

Sometimes you may want to full-manage the timing of performing screenshot.
Use the `waitFor` option if you think so. This string parameter should points a global function to return `Promise`.

For example, the following setting makes the screenshot function wait for firing of `fontLoading`:

```html
<!-- ./storybook/preview-head.html -->
<link rel="preload" href="/some-heavy-asset.woff" as="font" onload="this.setAttribute('loaded', 'loaded')">
<script>
function fontLoading() {
  const loaded = () => !!document.querySelector('link[rel="preload"][loaded="loaded"]');
  if (loaded()) return Promise.resolve();
  return new Promise((resolve, reject) => {
    const id = setInterval(() => {
      if (!loaded()) return;
      clearInterval(id);
      resolve();
    }, 50);
  });
}
</script>
```

```javascript
import { setScreenshotOptions } from 'storybook-chrome-screenshot';

setScreenshotOptions({
  waitFor: 'fontLoading',
});
```

## Examples

- [tsuyoshiwada/scs-with-reg-viz](https://github.com/tsuyoshiwada/scs-with-reg-viz) : A example repository of visual regression test using storybook-chrome-screenshot and reg-suit.
- [Quramy/angular-sss-demo](https://github.com/Quramy/angular-sss-demo) : Storybook, Screenshot, and Snapshot testing for Angular
- [viswiz-io/viswiz-tutorial-storybook](https://github.com/viswiz-io/viswiz-tutorial-storybook) : A tutorial repository for setting up visual regression testing with VisWiz.io

## TODO

The following tasks remain. Contributes are welcome :smiley:

- [x] Global Options.
- [x] ~~Shooting at an arbitrary timing.~~ (No plan for support)
- [x] Support for [Angular](https://angular.io).
- [x] Support for [Vue.js](https://github.com/vuejs/vue).
- [ ] More unit testing.

## Contribute

1.  Fork it!
1.  Create your feature branch: `git checkout -b my-new-feature`
1.  Commit your changes: `git commit -am 'Add some feature'`
1.  Push to the branch: `git push origin my-new-feature`
1.  Submit a pull request :muscle:

Bugs, feature requests and comments are more than welcome in the [issues](https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues).

### Development

We will develop using the following npm scripts.

#### `npm run test`

We will run Lint, unit test, E2E test in order.  
Each test can also be executed individually with the following command.

```bash
# Run TSLint
$ npm run test:lint

# Run unit test using Jest
$ npm run test:unit

# Run E2E test
$ npm run test:e2e

# Run unit tests in watch mode
$ npm run test:watch
```

#### `npm run build`

Compile the source code written in TypeScript.

## License

[MIT © tsuyoshiwada](./LICENSE)


CHANGELOG.md
[api]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/tree/develop#api
[#1]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/1
[#3]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/3
[#4]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/4
[#5]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/5
[#6]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/6
[#7]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/7
[#8]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/8
[#9]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/9
[#10]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/10
[#12]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/12
[#14]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/14
[#22]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/22
[#24]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/24
[#34]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/pull/34
[#35]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/pull/35
[#43]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/pull/43
[#46]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/pull/46
[#50]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/pull/50
[#52]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/pull/52
[#53]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/issues/53
[#54]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/pull/54
[#55]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/pull/55
[#73]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/pull/73
[#73]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/pull/73
[#84]: https://github.com/tsuyoshiwada/storybook-chrome-screenshot/pull/84

## 1.3.1

> 2018-11-08

### Bugfix

- Use the height that has been set in the viewport config for the screenshot ([#84][#84]). Thanks [@meghna](https://github.com/meghna)!

## 1.3.0

> 2018-08-05

### New features

- Full control the screenshot timing ([#73][#73]), Thanks [@Quramy](https://github.com/Quramy) !!

## 1.2.1

> 2018-07-28

### New features

- Bump `puppeteer` to 1.6.1

### Bugfix

- Fixed a bug where the bottom edge of the image can be overlooked.
- `@types/commander` has been removed from dependency.

### Development

- Update code format (use `tslint-microsoft-contrib`)
- Add prettier
- Add renovate app

## 1.2.0

> 2018-07-26

### New features

- Add file saving pattern ([#46][#46]), Thanks [@theKashey](https://github.com/theKashey) !

## 1.1.0

> 2018-06-22

### New features

- Support for Knobs integration :tada: ([#35][#35])
  - [@jessepinho](https://github.com/jessepinho), Thank you for feature proposal !!
  - **NOTE:** Angular is unsupported until Issue [here](https://github.com/storybooks/storybook/issues/3042) is resolved.
  - Knobs integration frameworks
    - React
    - Vue.js
- Add `--puppeteer-launch-config` CLI config ([#55][#55]), Thanks [@janpaul123](https://github.com/janpaul123) !!

### Performance Improves

- Parallelize browser processes rather than pages [#50][#50]. Thanks [@Quramy](https://github.com/Quramy) !!
  - The performance of `CAPTURE` phase has improved greatly.
- Perf tune on prepare phase [#52][#52]. Thanks [@Quramy](https://github.com/Quramy) !!!!
  - The performance of `PREPARE` phase has improved very greatly.

### Bugfix

- Fix parallel rendering [#43][#43]. Thanks [@theKashey](https://github.com/theKashey) !!
- Fixed a bug that caused an error when used in conjunction with `Storyshots`. Thanks [@Quramy](https://github.com/Quramy) !!

### Minor changes

- Parallelize shooting of chunked Stories.
- Update dependencies & devDependencies.

## 1.0.1

> 2018-03-14

### Minor changes

- Log when something is wrong. (Thanks [@kogai](https://github.com/kogai) !!)

## 1.0.0

> 2018-01-11

First major release :tada:

### New features

- Support for Vue.js :tada: :tada:

### Breaking changes

- Change interface of `initScreenshot()`.
  - Changed to function execution in order to avoid destructive change when it came to receive options as with withScreenshot in the future.

### Minor changes

- Change output format capture in CI environment.

## 0.10.0

> 2018-01-07

### Minor changes

- Switch to TypeScript from Babel.
- Update dependencies & devDependencies.
- package structure.
  - Refactoring.
  - It is easy to deal with various frameworks.
  - We plan support to Vue.js in the future.
- Add unit testing.
- Add E2E testing. [#24][#24] Thank you [@Quramy](https://github.com/Quramy) !!
- Fix document.
- Update output format.

## 0.9.0

> 2018-01-03

### Bugfix

- Fix puppeteer launch on Linux based CI. Thank you [@Quramy](https://github.com/Quramy) !!

## 0.9.0

> 2018-01-03

- Angular Support :tada: Thank you [@Quramy](https://github.com/Quramy) !

## 0.8.2

> 2017-12-29

### Bugfix

- Fix crash bug with storybook v3.3.x

## 0.8.1

> 2017-12-22

### Bugfix

- Broken storybook ... [#12][#12]
  - Revert feature `module` in package.json.

## 0.8.0

> 2017-12-10

### New features

- Add `initScreenshot` API. [#14][#14]

### Bugfix

- Fix capturing process in the big projects [#14][#14], [#10][#10]

### Big thanks

- [@alexeybondarenko](https://github.com/alexeybondarenko)

## 0.7.0

> 2017-12-06

### New features

- Add module entry to `package.json`. ([#12][#12])
  - Thank you [@marcobiedermann](https://github.com/marcobiedermann) !

## 0.6.1

> 2017-10-18

### Bugfix

- Fixed a bug that screenshot was not taken when using `addDecorator`. ([#10][#10])

## 0.6.0

> 2017-09-28

### New features

- Add parallel CLI options. ([#7][#7])
  - `--parallel`
  - By launching multiple Page instances of Puppeteer, you can expect a slight improvement in performance.
- Add injection scripts CLI options. ([#8][#8])

src/cli.ts
#!/usr/bin/env node

import 'babel-polyfill'; // tslint:disable-line: no-import-side-effect
import { execSync, spawn } from 'child_process';
import program from 'commander';
import isCI from 'is-ci';
import * as path from 'path';
import { App, Browser, Server, StoryStore, Terminal } from './core/app';
import { parser } from './core/utils';
import { CLIOptions } from './models/options';

// tslint:disable-next-line: no-var-requires no-require-imports
const pkg = require('../package.json');

// tslint:disable: max-line-length
program
  .version(pkg.version)
  .usage('[options]')
  .option('-p, --port [number]', 'Storybook server port.', parser.identity, 9001)
  .option('-h, --host [string]', 'Storybook server host.', parser.identity, 'localhost')
  .option('-s, --static-dir <dir-names>', 'Directory where to load static files from.', parser.list)
  .option(
    '-c, --config-dir [dir-name]',
    'Directory where to load Storybook configurations from.',
    parser.identity,
    '.storybook'
  )
  .option(
    '-o, --output-dir [dir-name]',
    'Directory where screenshot images are saved.',
    parser.identity,
    '__screenshots__'
  )
  .option(
    '--parallel [number]',
    'Number of Page Instances of Puppeteer to be activated when shooting screenshots.',
    parser.number,
    4
  )
  .option(
    '--filter-kind [regexp]',
    'Filter of kind with RegExp. (example: "Button$")',
    parser.regexp
  )
  .option(
    '--filter-story [regexp]',
    'Filter of story with RegExp. (example: "^with\\s.+$")',
    parser.regexp
  )
  .option(
    '--inject-files <file-names>',
    'Path to the JavaScript file to be injected into frame.',
    parser.list,
    []
  )
  .option(
    '--browser-timeout [number]',
    'Timeout milliseconds when Puppeteer opens Storybook.',
    parser.number,
    30000
  )
  .option(
    '--puppeteer-launch-config [json]',
    'JSON string of launch config for Puppeteer.',
    parser.identity,
    '{"args":["--no-sandbox","--disable-setuid-sandbox", "--disable-dev-shm-usage"]}'
  )
  .option('--silent', 'Suppress standard output.', parser.identity, false)
  .option('--debug', 'Enable debug mode.', parser.identity, false)
  .parse(process.argv);
// tslint:enable

const bin = execSync('echo $(npm bin)', { encoding: 'utf-8' })
  .toString()
  .trim();

const options: CLIOptions = {
  port: program.port,
  host: program.host,
  staticDir: program.staticDir,
  configDir: program.configDir,
  outputDir: program.outputDir,
  filterKind: program.filterKind,
  filterStory: program.filterStory,
  browserTimeout: program.browserTimeout,
  parallel: program.parallel,
  injectFiles: program.injectFiles,
  puppeteerLaunchConfig: program.puppeteerLaunchConfig,
  silent: !!program.silent,
  debug: !!program.debug,
  ciMode: isCI,
  cwd: path.resolve(bin, '..', '..'),
  cmd: path.resolve(bin, 'start-storybook')
};

// tslint:disable-next-line: no-floating-promises
(async () => {
  const store = new StoryStore(options.filterKind, options.filterStory);

  const terminal = new Terminal(
    process.stdout,
    process.stderr,
    options.silent,
    options.debug,
    options.ciMode
  );

  const app = new App(
    options,
    store,
    terminal,
    new Server(options, terminal, spawn),
    (id: number) => new Browser(id, store, options)
  );

  try {
    await app.validate();
    await app.launch();
    await app.prepare();
    await app.capture();
    await app.teardown();
    process.exit(0);
  } catch (e) {
    await app.terminate(e);
    process.exit(1);
  }
})();
